cpp/source/deal.II/mp_values.templates.h

```diff
@@ -6,82 +6,273 @@
+        MPValuesParameters<dim, spacedim> const &params)
+{
+  // Construct a random number generator and use the MPI rank as a seed.
+  std::default_random_engine generator(
+      params.geometry->get_mpi_communicator().rank());
 
+  boost::property_tree::ptree const &database = *params.database;
+  // Build a map material_id -> material_name
+  std::map<dealii::types::material_id, std::string> material_map;
+  for (auto const &m : *params.geometry->get_materials())
+  {
+    std::string const &material_name = m.first;
+    // Geometry(...) tags the current collectors with "current_collector" so
+    // we ignore these.
+    if (material_name.compare("collector_anode") == 0 ||
+        material_name.compare("collector_cathode") == 0)
+      continue;
+    std::vector<dealii::types::material_id> const &material_ids = m.second;
+    boost::property_tree::ptree const &material_database =
+        database.get_child(material_name);
+    std::string const type = material_database.get<std::string>("type");
+    for (auto const &id : material_ids)
+    {
+      auto ret = material_map.emplace(id, material_name);
+      if (!ret.second)
+        throw std::runtime_error("material id " + std::to_string(id) +
+                                 " assigned multiple times");
+    }
+  }
+  auto const parameters = database.get<int>("parameters");
+  // Build a map material_id -> parameters
+  // TODO: For simplicity let's perturb all parameters for now
+  // The map is only a map parameter path in the ptree -> object that emulates
+  // the std random distribution (i.e. takes a generator object as argument and
+  // returns a double)
+  std::map<std::string, std::function<double(std::default_random_engine &)>>
+      parameter_map;
+  for (int p = 0; p < parameters; ++p)
+  {
+    boost::property_tree::ptree const &parameter_database =
+        database.get_child("parameter_" + std::to_string(p));
+    auto const parameter_path = parameter_database.get<std::string>("path");
+    // Check that the parameter does exist
+    if (!database.get_optional<double>(parameter_path))
+      throw std::runtime_error("Parameter path " + parameter_path +
+                               " does not exist in the material database");
+    auto ret = parameter_map.emplace(parameter_path,
+                                     _build_parameter(parameter_database));
+    if (!ret.second)
+      throw std::runtime_error("Parameter " + parameter_path +
+                               "  is present multiple times");
+  }
+  // traverse the triangulation and build material properties with perturbed
+  // parameters in each cell
+  auto const &triangulation = *params.geometry->get_triangulation();
+  // Make a copy of the material database
+  auto dummy_database =
+      std::make_shared<boost::property_tree::ptree>(*params.database);
+  // Use a dummy version of the parameters to build the MPValues that points to
+  // the copy of the database.
```

Comment:
Can you explain why you need a dummy database and a dummy parameters. Or find better names. By dummy do you mean local to a cell?